src/Service/Service.php

```diff
@@ -0,0 +1,227 @@
+            }, []),
+            'started' => $this->started,
+        ];
+    }
+
+    public function addGroup(string $name): ServiceGroup
+    {
+        if (!array_key_exists($name, $this->groups)) {
+            $this->groups[$name] = new ServiceGroup($this, $name);
+        }
+
+        return $this->groups[$name];
+    }
+
+    public function addEndpoint(
+        string $name,
+        string|EndpointHandler|callable $endpointHandler,
+        array $options = []
+    ): void {
+        $subject = $name;
+        $queue_group = 'q';
+
+        if (array_key_exists('subject', $options)) {
+            $subject = $options['subject'];
+        }
+
+        if (array_key_exists('queue_group', $options)) {
+            $queue_group = $options['queue_group'];
+        }
+
+        if (array_key_exists($name, $this->endpoints)) {
+            throw new \LogicException("Endpoint $name already is defined");
+        }
+
+        $this->endpoints[$name] = new ServiceEndpoint(
+            $this,
+            $name,
+            $subject,
+            $endpointHandler,
+            $queue_group
+        );
+    }
+
+    public function reset(): void
+    {
+        array_map(
+            function (ServiceEndpoint $endpoint) {
+                $endpoint->resetStats();
+            },
+            $this->endpoints
+        );
+    }
+
+    private function registerVerbs(): void
+    {
+        $verbs = [
+            'PING' => function (Payload $payload) {
+                return $this->ping();
+            },
+            'INFO' => function (Payload $payload) {
+                return $this->info();
+            },
+            'STATS' => function (Payload $payload) {
+                return $this->stats();
+            },
+        ];
+
+        foreach ($verbs as $verb => $handler) {
+            // Add the all handler
+            $this->addInternalHandler($verb, '', '', "$verb-all", $handler);
+
+            // Add the kind handler
+            $this->addInternalHandler($verb, $this->name, '', "$verb-kind", $handler);
+
+            // Add the service id handler
+            $this->addInternalHandler($verb, $this->name, $this->id, $verb, $handler);
+        }
+    }
+
+    private function addInternalHandler(
+        string $verb,
+        string $kind,
+        string $id,
+        string $name,
+        callable $handler
+    ): void {
+        $subject = $this->controlSubject($verb, $kind, $id);
+
+        $this->subscriptions[$name] = $this->client->subscribe(
+            $subject,
+            $handler
+        );
+    }
+
+    private function controlSubject(string $verb, string $name, string $id): string
+    {
+        if ($name == '' && $id == '') {
+            return "\$SRV.$verb";
+        }
+
+        if ($id == '') {
+            return "\$SRV.$verb.$name";
+        }
+
+        return "\$SRV.$verb.$name.$id";
+    }
+
+    public function run() : void
+    {
+        print_r("$this->name is ready to accept connections\n");
```

Comment:
standard way to add debug information should be optional client.logger usage when present.
see `Connection.php` source for an example